compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -287,9 +252,10 @@ impl<'tcx> InliningMap<'tcx> {
         self.targets.reserve(new_items_count);
         self.inlines.ensure(new_items_count_total);
 
-        for (i, (Spanned { node: mono_item, .. }, inlined)) in new_targets.into_iter().enumerate() {
+        for (i, Spanned { node: mono_item, .. }) in new_targets.into_iter().enumerate() {
             self.targets.push(*mono_item);
-            if *inlined {
+            let is_inlined = mono_item.instantiation_mode(tcx) == InstantiationMode::LocalCopy;
```

Comment:
The key lies in the cost of calling `instantiation_mode`. Calling it within the lock will increase the locking time. It's not clear to me that this has a noticeable effect on the efficiency of monomorphization. cc @Zoxc

Reply:
The comment was added in #67756. Reviewers asked for more explanation in comments, but the PR author refused. It seems like it was for a deadlock. I ran the `ui` tests with the parallel front-end enabled without problem.

Reply:
#112128 is an alternative approach that avoids the whole locking issue.